src/mongo/unittest/unittest.cpp

```diff
@@ -386,6 +447,150 @@ int Suite::run(const std::vector<std::string>& suites, const std::string& filter
+        }
+    }
+
+    std::vector<std::string> torun(suites);
+
+    if (torun.empty()) {
+        for (SuiteMap::const_iterator i = _allSuites().begin(); i != _allSuites().end(); ++i) {
+            torun.push_back(i->first);
+        }
+    }
+
+    std::vector<Result*> results;
+
+    for (std::vector<std::string>::iterator i = torun.begin(); i != torun.end(); i++) {
+
+        std::string name = *i;
+        std::shared_ptr<Suite>& s = _allSuites()[name];
+        fassert(40143, s != NULL);
+
+        log() << "going to run suite: " << name << std::endl;
+        results.push_back(s->run(filter, runsPerTest));
+    }
+
+    log() << "**************************************************" << std::endl;
+
+    int rc = 0;
+
+    int tests = 0;
+    int asserts = 0;
+    int millis = 0;
+
+    Result totals("TOTALS");
+    std::vector<std::string> failedSuites;
+
+    Result::cur = NULL;
+    for (std::vector<Result*>::iterator i = results.begin(); i != results.end(); i++) {
+        Result* r = *i;
+        log() << r->toString();
+        if (abs(r->rc()) > abs(rc))
+            rc = r->rc();
+
+        tests += r->_tests;
+        if (!r->_fails.empty()) {
+            failedSuites.push_back(r->toString());
+            for (std::vector<std::string>::const_iterator j = r->_fails.begin();
+                 j != r->_fails.end();
+                 j++) {
+                const std::string& s = (*j);
+                totals._fails.push_back(r->_name + "/" + s);
+            }
+        }
+        asserts += r->_asserts;
+        millis += r->_millis;
+
+        delete r;
+    }
+
+    totals._tests = tests;
+    totals._asserts = asserts;
+    totals._millis = millis;
+
+    log() << totals.toString();  // includes endl
+
+    // summary
+    if (!totals._fails.empty()) {
+        log() << "Failing tests:" << std::endl;
+        for (std::vector<std::string>::const_iterator i = totals._fails.begin();
+             i != totals._fails.end();
+             i++) {
+            const std::string& s = (*i);
+            log() << "\t " << s << " Failed";
+        }
+        log() << "FAILURE - " << totals._fails.size() << " tests in " << failedSuites.size()
+              << " suites failed";
+    } else {
+        log() << "SUCCESS - All tests in all suites passed";
+    }
+
+    // Start building the JSON output using BSONObjBuilder
+    BSONObjBuilder b;
+
+    // TODO: Get the start and end time
+    b.append("start", "PLACEHOLDER");
+    b.append("end", "PLACEHOLDER");
+
+    // No errors
+    BSONArrayBuilder errorsList(b.subarrayStart("errors"));
+    errorsList.doneFast();
+
+    BSONArrayBuilder resultsList(b.subarrayStart("results"));
+    // Construct the results list by pulling the data from the static BenchmarkStats struct
+    for (BenchmarkStats stat : statsList) {
+        BSONObjBuilder statObj;
+        statObj.append("name", stat.name);
+        BSONObjBuilder benchResults(statObj.subobjStart("results"));
```

Comment:
It is helpful to create a nested scope when building inner BSONObjects. It makes the shape of the code match the shape of the Object.